File: backend/app/a2ui/json_pointer.py
from __future__ import annotations

from typing import Any


class JsonPointerError(ValueError):
    pass


def _unescape(token: str) -> str:
    return token.replace("~1", "/").replace("~0", "~")


def parse(pointer: str) -> list[str]:
    if pointer == "":
        return []
    if not pointer.startswith("/"):
        raise JsonPointerError(f"Invalid JSON Pointer: {pointer}")
    return [_unescape(part) for part in pointer.split("/")[1:]]


def get(doc: Any, pointer: str, default: Any = None) -> Any:
    if pointer in ("", "/"):
        return doc
    try:
        tokens = parse(pointer)
    except JsonPointerError:
        raise
    cur = doc
    for token in tokens:
        if isinstance(cur, list):
            if not token.isdigit():
                return default
            idx = int(token)
            if idx >= len(cur):
                return default
            cur = cur[idx]
        elif isinstance(cur, dict):
            if token not in cur:
                return default
            cur = cur[token]
        else:
            return default
    return cur


def set_value(doc: Any, pointer: str, value: Any) -> Any:
    if pointer in ("", "/"):
        if value is None:
            return {}
        return value
    tokens = parse(pointer)
    if not isinstance(doc, dict) and pointer != "/":
        doc = {} if doc is None else doc
    if not isinstance(doc, dict):
        raise JsonPointerError(f"Invalid JSON Pointer: {pointer}")
    cur: Any = doc
    for i, token in enumerate(tokens[:-1]):
        nxt = tokens[i + 1]
        is_index = nxt.isdigit()
        if isinstance(cur, list):
            idx = int(token)
            while len(cur) <= idx:
                cur.append({} if not is_index else [])
            cur = cur[idx]
        else:
            if token not in cur or cur[token] is None:
                cur[token] = [] if is_index else {}
            cur = cur[token]
    last = tokens[-1]
    if value is None:
        if isinstance(cur, list) and last.isdigit():
            idx = int(last)
            if 0 <= idx < len(cur):
                cur.pop(idx)
        elif isinstance(cur, dict):
            cur.pop(last, None)
        return doc
    if isinstance(cur, list) and last.isdigit():
        idx = int(last)
        while len(cur) <= idx:
            cur.append(None)
        cur[idx] = value
    elif isinstance(cur, dict):
        cur[last] = value
    else:
        raise JsonPointerError(f"Invalid JSON Pointer: {pointer}")
    return doc

File: frontend/src/a2ui/jsonPointer.ts
export class JsonPointerError extends Error {
  constructor(pointer: string) {
    super(`Invalid JSON Pointer: ${pointer}`);
    this.name = "JsonPointerError";
  }
}

function unescape(token: string) {
  return token.replace(/~1/g, "/").replace(/~0/g, "~");
}

export function parsePointer(pointer: string): string[] {
  if (pointer === "") return [];
  if (!pointer.startsWith("/")) throw new JsonPointerError(pointer);
  return pointer.split("/").slice(1).map(unescape);
}

export function getPointer(doc: unknown, pointer: string): unknown {
  if (pointer === "" || pointer === "/") return doc;
  const tokens = parsePointer(pointer);
  let cur: unknown = doc;
  for (const token of tokens) {
    if (token === "@index") return undefined;
    if (Array.isArray(cur)) {
      const idx = Number(token);
      if (!Number.isInteger(idx) || idx >= cur.length) return undefined;
      cur = cur[idx];
    } else if (cur && typeof cur === "object") {
      cur = (cur as Record<string, unknown>)[token];
    } else {
      return undefined;
    }
  }
  return cur;
}

export function setPointer(doc: unknown, pointer: string, value: unknown): unknown {
  if (pointer === "" || pointer === "/") return value ?? {};
  const tokens = parsePointer(pointer);
  const root = doc && typeof doc === "object" ? (doc as Record<string, unknown>) : {};
  let cur: any = root;
  for (let i = 0; i < tokens.length - 1; i++) {
    const token = tokens[i];
    const nextIsIndex = /^\d+$/.test(tokens[i + 1]);
    if (Array.isArray(cur)) {
      const idx = Number(token);
      if (!cur[idx]) cur[idx] = nextIsIndex ? [] : {};
      cur = cur[idx];
    } else {
      if (cur[token] == null) cur[token] = nextIsIndex ? [] : {};
      cur = cur[token];
    }
  }
  const last = tokens[tokens.length - 1];
  if (value === null) {
    if (Array.isArray(cur) && /^\d+$/.test(last)) cur.splice(Number(last), 1);
    else if (cur && typeof cur === "object") delete cur[last];
    return root;
  }
  if (Array.isArray(cur) && /^\d+$/.test(last)) cur[Number(last)] = value;
  else cur[last] = value;
  return root;
}
